File: path-sum.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x, left=None, right=None):
        self.val = x
        self.left = left
        self.right = right


class Solution(object):
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        if root is None:
            return False

        def dfs(root, s=0):
            if root.left is None and root.right is None:
                if s == sum:
                    dfs.flag = True
                return
            if root and root.left:
                dfs(root.left, root.left.val + s)
            if root and root.right:
                dfs(root.right, root.right.val + s)
        dfs.flag = False
        dfs(root, root.val)
        return dfs.flag


def test():
    t = TreeNode(5,
                 TreeNode(4, TreeNode(11, TreeNode(7), TreeNode(2))),
                 TreeNode(8, TreeNode(13), TreeNode(4, None, TreeNode(1))))
    s = Solution()
    assert True == s.hasPathSum(t, 22)

if __name__ == '__main__':
    test()

File: algorithms/easy/path-sum.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>


struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct TreeNode *newNode(int val, struct TreeNode *left, struct TreeNode *right) {
    struct TreeNode *node = malloc(sizeof(struct TreeNode));
    node->val = val;
    node->left = left;
    node->right = right;
    return node;
}


bool _hasPathSum(struct TreeNode *root, int curSum, int sum) {
    if (root != NULL && root->left == NULL && root->right == NULL && curSum + root->val == sum)
        return true;
    if (root->left != NULL && _hasPathSum(root->left, curSum + root->val, sum))
        return true;
    if (root->right != NULL && _hasPathSum(root->right, curSum + root->val, sum))
        return true;
    return false;
}


bool hasPathSum(struct TreeNode* root, int sum){
    if (root == NULL)
        return false;
    return _hasPathSum(root, 0, sum);
}

int main() {
    struct TreeNode *n15 = newNode(15, NULL, NULL);
    struct TreeNode *n7 = newNode(7, NULL, NULL);
    struct TreeNode *n20 = newNode(20, n15, n7);
    struct TreeNode *n9 = newNode(9, NULL, NULL);
    struct TreeNode *n3 = newNode(3, n9, n20);
    printf("%d\n", hasPathSum(n3, 12));
}
